From e537acdf7a5efcd465951c70ccde9ecbe29a927b Mon Sep 17 00:00:00 2001
From: c4pp4
Date: Wed, 21 Oct 2026 20:55:24 +0200
Subject: [PATCH 1/1] Unity shortcuts plugin

Control the player with the multimedia keys in Unity. Based on the GNOME
Shortcuts plugin removed upstream in audacious-plugins 3.10 (commit
3dc8222f59, "GNOME 3.26 broke compatibility with this plugin").

unity-settings-daemon has no MPRIS support for media keys, it forwards
them only to applications registered via GrabMediaPlayerKeys on
org.gnome.SettingsDaemon.MediaKeys by the MediaPlayerKeyPressed signal.

Handled keys: Play/Pause, Stop, Previous, Next, Rewind and FastForward
(seek by the "step_size" setting). Volume and mute keys are handled by
unity-settings-daemon itself and never forwarded. Repeat and Shuffle
are not handled, their keycodes are above 255 and not reachable in X11.

Signed-off-by: c4pp4
---
 meson.build                          |   1 +
 meson_options.txt                    |   2 +
 src/meson.build                      |   4 +
 src/unityshortcuts/meson.build       |  10 ++
 src/unityshortcuts/unityshortcuts.cc | 229 +++++++++++++++++++++++++++
 5 files changed, 246 insertions(+)
 create mode 100644 src/unityshortcuts/meson.build
 create mode 100644 src/unityshortcuts/unityshortcuts.cc

diff --git a/meson.build b/meson.build
index 8260aef..5ef903d 100644
--- a/meson.build
+++ b/meson.build
@@ -316,6 +316,7 @@ summary({
   'MPRIS 2 Server': get_variable('have_mpris2', false),
   'Scrobbler 2.0': get_variable('have_scrobbler2', false),
   'Song Change': get_option('songchange'),
+  'Unity Shortcuts': get_option('unityshortcuts'),
 }, section: 'General')
 
 if conf.has('USE_QT')
diff --git a/meson_options.txt b/meson_options.txt
index a3e5ba1..a1c4afc 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -155,6 +155,8 @@ option('statusicon', type: 'boolean', value: true,
        description: 'Whether the Status Icon plugin is enabled')
 option('streamtuner', type: 'boolean', value: false,
        description: 'Whether the Stream Tuner plugin is enabled')
+option('unityshortcuts', type: 'boolean', value: true,
+       description: 'Whether the Unity Shortcuts plugin is enabled')
 
 
 # effect plugins
diff --git a/src/meson.build b/src/meson.build
index 34b48ce..8fdd12f 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -282,6 +282,10 @@ if get_option('songchange')
   subdir('songchange')
 endif
 
+if get_option('unityshortcuts')
+  subdir('unityshortcuts')
+endif
+
 
 # input plugins
 if get_option('aac')
diff --git a/src/unityshortcuts/meson.build b/src/unityshortcuts/meson.build
new file mode 100644
index 0000000..5ec61ee
--- /dev/null
+++ b/src/unityshortcuts/meson.build
@@ -0,0 +1,10 @@
+dbus_dep = dependency('dbus-1', version: '>= 0.60', required: true)
+dbus_glib_dep = dependency('dbus-glib-1', version: '>= 0.60', required: true)
+
+shared_module('unityshortcuts',
+  'unityshortcuts.cc',
+  dependencies: [audacious_dep, gtk_dep, dbus_dep, dbus_glib_dep],
+  name_prefix: '',
+  install: true,
+  install_dir: general_plugin_dir
+)
diff --git a/src/unityshortcuts/unityshortcuts.cc b/src/unityshortcuts/unityshortcuts.cc
new file mode 100644
index 0000000..958caca
--- /dev/null
+++ b/src/unityshortcuts/unityshortcuts.cc
@@ -0,0 +1,229 @@
+/*
+ *  This file is part of audacious-unity-shortcut plugin for audacious
+ *
+ *  Copyright (c) 2007-2008    Sascha Hlusiak <contact@saschahlusiak.de>
+ *  Name: plugin.c
+ *  Description: plugin.c
+ *
+ *  audacious-unity-shortcut is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  audacious-unity-shortcut is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with audacious-unity-shortcut; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include <string.h>
+#include <dbus/dbus.h>
+#include <dbus/dbus-glib-bindings.h>
+#include <glib-object.h>
+
+#include <libaudcore/drct.h>
+#include <libaudcore/plugin.h>
+#include <libaudcore/i18n.h>
+#include <libaudcore/runtime.h>
+
+class UnityShortcuts : public GeneralPlugin
+{
+public:
+    static const char about[];
+
+    static constexpr PluginInfo info = {
+        N_("Unity Shortcuts"),
+        PACKAGE,
+        about
+    };
+
+    constexpr UnityShortcuts () : GeneralPlugin (info, true) {}
+
+    bool init ();
+    void cleanup ();
+};
+
+EXPORT UnityShortcuts aud_plugin_instance;
+
+static DBusGProxy *media_player_keys_proxy = nullptr;
+
+const char UnityShortcuts::about[] =
+ N_("Unity Shortcuts Plugin\n"
+    "(former GNOME Shortcuts Plugin)\n"
+    "Lets you control the player with Unity's shortcuts.\n\n"
+    "Copyright (C) 2007-2008 Sascha Hlusiak <contact@saschahlusiak.de>");
+
+#define g_marshal_value_peek_string(v)   (char*) g_value_get_string (v)
+
+
+static void
+hotkey_marshal_VOID__STRING_STRING (GClosure     *closure,
+                                    GValue       *return_value,
+                                    unsigned         n_param_values,
+                                    const GValue *param_values,
+                                    void *      invocation_hint,
+                                    void *      marshal_data)
+{
+    typedef void (*GMarshalFunc_VOID__STRING_STRING) (void * data1,
+                                                      void * arg_1,
+                                                      void * arg_2);
+    GMarshalFunc_VOID__STRING_STRING callback;
+    GCClosure *cc = (GCClosure*) closure;
+    void * data1;
+
+    g_return_if_fail (n_param_values == 3);
+
+    if (G_CCLOSURE_SWAP_DATA (closure))
+    {
+        data1 = closure->data;
+    } else {
+        data1 = g_value_peek_pointer (param_values + 0);
+    }
+    callback = (GMarshalFunc_VOID__STRING_STRING) (marshal_data ? marshal_data : cc->callback);
+
+    callback (data1,
+     g_marshal_value_peek_string (param_values + 1),
+     g_marshal_value_peek_string (param_values + 2));
+}
+
+static void
+on_media_player_key_pressed (DBusGProxy *proxy, const char *application, const char *key)
+{
+    if (strcmp ("Audacious", application) == 0) {
+        /* play or pause */
+        if (strcmp ("Play", key) == 0 || strcmp ("Pause", key) == 0)
+        {
+            aud_drct_play_pause ();
+            return;
+        }
+
+        /* stop */
+        if (strcmp ("Stop", key) == 0)
+        {
+            aud_drct_stop ();
+            return;
+        }
+
+        /* prev track */
+        if (strcmp ("Previous", key) == 0)
+        {
+            aud_drct_pl_prev ();
+            return;
+        }
+
+        /* next track */
+        if (strcmp ("Next", key) == 0)
+        {
+            aud_drct_pl_next ();
+            return;
+        }
+
+        /* rewind */
+        if (strcmp ("Rewind", key) == 0)
+        {
+            aud_drct_seek (aud_drct_get_time () - aud_get_int ("step_size") * 1000);
+            return;
+        }
+
+        /* fast forward */
+        if (strcmp ("FastForward", key) == 0)
+        {
+            aud_drct_seek (aud_drct_get_time () + aud_get_int ("step_size") * 1000);
+            return;
+        }
+    }
+}
+
+void UnityShortcuts::cleanup ()
+{
+    GError *error = nullptr;
+    if (media_player_keys_proxy == nullptr) return;
+
+    dbus_g_proxy_disconnect_signal (media_player_keys_proxy, "MediaPlayerKeyPressed",
+     G_CALLBACK (on_media_player_key_pressed), nullptr);
+
+    dbus_g_proxy_call (media_player_keys_proxy,
+     "ReleaseMediaPlayerKeys", &error,
+     G_TYPE_STRING, "Audacious",
+     G_TYPE_INVALID, G_TYPE_INVALID);
+    if (error != nullptr) {
+        g_warning ("Could not release media player keys: %s", error->message);
+        g_error_free (error);
+    }
+    g_object_unref(media_player_keys_proxy);
+    media_player_keys_proxy = nullptr;
+}
+
+bool UnityShortcuts::init ()
+{
+    DBusGConnection *bus;
+    GError *error = nullptr;
+    dbus_g_thread_init();
+
+    bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);
+    if ((bus == nullptr) && error) {
+        g_warning ("Error connecting to DBus: %s", error->message);
+        g_error_free (error);
+        return false;
+    } else {
+        media_player_keys_proxy = dbus_g_proxy_new_for_name (bus,
+         "org.gnome.SettingsDaemon",
+         "/org/gnome/SettingsDaemon/MediaKeys",
+         "org.gnome.SettingsDaemon.MediaKeys");
+        if (media_player_keys_proxy == nullptr) {
+            media_player_keys_proxy = dbus_g_proxy_new_for_name (bus,
+             "org.gnome.SettingsDaemon.MediaKeys",
+             "/org/gnome/SettingsDaemon/MediaKeys",
+             "org.gnome.SettingsDaemon.MediaKeys");
+            if (media_player_keys_proxy == nullptr)
+                return false;
+        }
+
+        dbus_g_proxy_call (media_player_keys_proxy,
+         "GrabMediaPlayerKeys", &error,
+         G_TYPE_STRING, "Audacious",
+         G_TYPE_UINT, 0,
+         G_TYPE_INVALID,
+         G_TYPE_INVALID);
+        if (error != nullptr) {
+            g_error_free (error);
+            error = nullptr;
+            g_object_unref(media_player_keys_proxy);
+            media_player_keys_proxy = dbus_g_proxy_new_for_name (bus,
+             "org.gnome.SettingsDaemon",
+             "/org/gnome/SettingsDaemon",
+             "org.gnome.SettingsDaemon");
+            if (media_player_keys_proxy == nullptr)
+                return false;
+
+            dbus_g_proxy_call (media_player_keys_proxy,
+             "GrabMediaPlayerKeys", &error,
+             G_TYPE_STRING, "Audacious",
+             G_TYPE_UINT, 0,
+             G_TYPE_INVALID,
+             G_TYPE_INVALID);
+            if (error != nullptr) {
+                g_warning ("Could not grab media player keys: %s", error->message);
+                g_error_free (error);
+                g_object_unref(media_player_keys_proxy);
+                media_player_keys_proxy = nullptr;
+                return false;
+            }
+        }
+
+        dbus_g_object_register_marshaller (hotkey_marshal_VOID__STRING_STRING,
+         G_TYPE_NONE, G_TYPE_STRING, G_TYPE_STRING, G_TYPE_INVALID);
+
+        dbus_g_proxy_add_signal (media_player_keys_proxy, "MediaPlayerKeyPressed",
+         G_TYPE_STRING, G_TYPE_STRING, G_TYPE_INVALID);
+
+        dbus_g_proxy_connect_signal (media_player_keys_proxy, "MediaPlayerKeyPressed",
+         G_CALLBACK (on_media_player_key_pressed), nullptr, nullptr);
+
+        return true;
+    }
+}
-- 
2.55.0

